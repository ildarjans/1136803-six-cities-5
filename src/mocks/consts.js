export const MOCK_OFFERS_COUNT = 100;
export const MOCK_REVIEWS_COUNT = 5;
export const AVATAR_URL = `https://avatars.dicebear.com/api`;
export const MAX_FEATURES = 3;
export const imageGallery = {
  MIN: 2,
  MAX: 8,
};

export const priceRange = {
  MIN: 19,
  MAX: 4999,
};
export const bedroomsRange = {
  MIN: 1,
  MAX: 6,
};
export const guestRange = {
  MIN: 1,
  MAX: 10,
};

export const avatarSprites = [
  `male`,
  `female`,
  `human`,
  `bottts`,
  `avataaars`,
  `gridy`,
];


export const names = [
  `James`,
  `John`,
  `Robert`,
  `William`,
  `Joseph`,
  `Charles`,
  `Thomas`,
  `Christopher`,
  `Anthony`,
  `Mark`,
  `Paul`,
  `Steven`,
  `Kenneth`,
  `Edward`,
  `Kevin`,
  `Ronald`,
  `Jason`,
  `Jeffrey`,
  `Gary`,
  `Nicholas`,
  `Stephen`,
  `Jonathan`,
  `Frank`,
  `Justin`,
  `Gregory`,
  `Samuel`,
  `Mary`,
  `Patricia`,
  `Elizabeth`,
  `Jessica`,
  `Sarah`,
  `Dorothy`,
  `Karen`,
  `Nancy`,
  `Sandra`,
  `Jennifer`,
  `Amanda`,
  `Rebecca`,
  `Sharon`,
  `Cynthia`,
  `Pamela`,
  `Katherine`,
  `Nicole`,
  `Samantha`,
  `Carolyn`,
];
export const lastNames = [
  `Smith`,
  `Brown`,
  `Johnson`,
  `Jones`,
  `Williams`,
  `Davis`,
  `Clark`,
  `White`,
  `Moore`,
  `Thompson`,
  `Adams`,
  `Thomas`,
  `Anderson`,
  `Lewis`,
  `Jackson`,
  `Green`,
  `Wood`,
  `Parker`,
  `Campbell`,
  `Robinson`,
  `Scott`,
  `Phillips`,
  `Turner`,
  `Carter`,
];

export const features = [
  `Wi-fi`,
  `Dishwasher`,
  `Parking`,
  `Washer`,
  `Elevator`,
  `Conditioner`,
  `Gym`,
  `Guide`,
  `Video games`,
  `Towels`,
  `Kitchen`,
  `Heating`,
];

export const descriptions = [
  `Lorem ipsum dolor sit amet, consectetur adipiscingâ€¦empor incididunt ut labore et dolore magna aliqua`,
  `Odio aenean sed adipiscing diam donec adipiscing tristique risus nec`,
  `Ut ornare lectus sit amet est placerat`,
  `Habitant morbi tristique senectus et`,
  `Ac tortor vitae purus faucibus ornare suspendisse sed nisi lacus`,
  `Tortor aliquam nulla facilisi cras fermentum odio eu feugiat pretium`,
  `Mattis enim ut tellus elementum`,
  `Justo donec enim diam vulputate ut pharetra sit amet aliquam`,
  `Elit ut aliquam purus sit amet luctus venenatis lectus`,
  `Eget velit aliquet sagittis id`,
  `Mi sit amet mauris commodo quis imperdiet massa`,
  `Ullamcorper a lacus vestibulum sed arcu non odio`,
  `Elementum eu facilisis sed odio morbi quis commodo odio`,
  `Et malesuada fames ac turpis egestas integer eget aliquet nibh`,
  `Pharetra massa massa ultricies mi quis hendrerit dolor`,
  `Eu turpis egestas pretium aenean pharetra magna ac`,
  `Dictum fusce ut placerat orci`,
  `Nisi vitae suscipit tellus mauris a diam maecenas sed`,
  `Mauris sit amet massa vitae tortor condimentum lacinia quis vel`,
  `Id faucibus nisl tincidunt eget nullam non nisi est sit.`,
];

export const titles = [
  `Beautiful & luxurious apartment at great location`,
  `Couples particularly like this location `,
  `Big company apartment`,
  `Nearest airport & transport station`,
  `Quite place for zen vacations`,
  `Best lobby bar & Michelin restaurant`,
  `Bed and breakfast. Best offer!`,
];
