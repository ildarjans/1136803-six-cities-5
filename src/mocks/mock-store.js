import configureStore from "redux-mock-store";
import {
  mockCityOffers,
  mockFavoritesOffers,
  mockOffers,
  mockReviews,
  mockUser
} from "./mock-data";

const createStore = configureStore();

const mockStoreTemplate = {
  OFFERS: {
    offers: mockOffers,
    cityOffers: mockCityOffers,
    loading: true,
    error: null
  },
  REVIEWS: {
    reviews: mockReviews,
    loading: false,
    error: null,
    posting: false
  },
  NEAR_OFFERS: {
    nearOffers: mockCityOffers.AMSTERDAM.slice(0, 3),
    loading: false,
    error: null
  },
  PROCESS: {
    activeCity: `AMSTERDAM`,
    hoveredOfferId: ``,
    sortType: ``
  },
  USER: {
    authorizationStatus: `NO_AUTHORIZED`,
    waitingAuthorizationResponse: false,
    error: null,
    user: mockUser
  },
  FAVORITES: {
    favoriteCityOffers: mockFavoritesOffers,
    updating: false,
    loading: false,
    error: null,
  }
};

export const mockStore = createStore(mockStoreTemplate);
