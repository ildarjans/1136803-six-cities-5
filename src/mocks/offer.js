import {
  generateId,
  getLocationCoords,
  getRandomArrayElement,
  getRandomArrayElements,
  getRandomAvatar,
  getRandomFullName,
  getRandomImageGallery,
  getRandomInteger,
} from "./utils";
import {offerTypes, cities} from "../const";
import {
  descriptions,
  features,
  BEDROOMS_RANGE,
  GUEST_RANGE,
  MAX_FEATURES,
  PRICE_RANGE,
  titles,
} from "./consts";

export function getOffer() {
  const city = getRandomArrayElement(cities);
  return {
    city,
    id: generateId(),
    images: getRandomImageGallery(),
    premium: Boolean(getRandomInteger(1)),
    price: getRandomInteger(PRICE_RANGE.MAX, PRICE_RANGE.MIN),
    title: getRandomArrayElement(titles),
    type: getRandomArrayElement(offerTypes),
    rating: getRandomInteger(5),
    description: getRandomArrayElement(descriptions),
    bedrooms: getRandomInteger(BEDROOMS_RANGE.MAX, BEDROOMS_RANGE.MIN),
    guests: getRandomInteger(GUEST_RANGE.MAX, GUEST_RANGE.MIN),
    features: getRandomArrayElements(features, MAX_FEATURES),
    owner: {
      avatar: getRandomAvatar(),
      name: getRandomFullName(),
      super: Boolean(getRandomInteger(1)),
    },
    coords: getLocationCoords(city)
  };
}

