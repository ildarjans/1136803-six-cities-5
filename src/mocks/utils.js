import {
  AVATAR_URL,
  AVATAR_SPRITES,
  LAST_NAMES,
  NAMES,
  IMAGE_GALLERY,
} from "./consts";

export function getRandomInteger(max, min = 0) {
  return min + Math.floor(Math.random() * (max + 1 - min));
}

export function getRandomArrayElements(arr, count) {
  if (arr.length < count) {
    return undefined;
  }
  const resultArray = [];
  while (resultArray.length < count) {
    const randElement = arr[getRandomInteger(arr.length - 1)];
    if (!resultArray.includes(randElement)) {
      resultArray.push(randElement);
    }
  }
  return resultArray;
}

export function getRandomArrayElement(arr) {
  if (!arr && arr.length === 0) {
    return undefined;
  }
  return arr[getRandomInteger(arr.length - 1)];
}

export function getRandomFullName() {
  return (
    `${getRandomArrayElement(NAMES)} ${getRandomArrayElement(LAST_NAMES)}`
  );
}

export function getRandomAvatar() {
  return `${AVATAR_URL}/${getRandomArrayElement(AVATAR_SPRITES)}/${Math.random()}.svg`;
}

export function getRandomImageGallery() {
  const gallerySize = getRandomInteger(IMAGE_GALLERY.MAX, IMAGE_GALLERY.MIN);
  return Array(gallerySize).fill().map(getRandomPhoto);
}

function getRandomPhoto() {
  return `http://picsum.photos/248/152?r=${Math.random()}`;
}

export function generateId() {
  return getRandomInteger(Date.now()).toString(16);
}
