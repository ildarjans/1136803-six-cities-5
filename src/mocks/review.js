import {
  getRandomArrayElement,
  getRandomAvatar,
  getRandomFullName,
  getRandomInteger,
} from "./utils";
import {descriptions} from "./consts";

export function getReview() {
  return {
    avatar: getRandomAvatar(),
    name: getRandomFullName(),
    date: new Date(),
    rating: getRandomInteger(5),
    review: getRandomArrayElement(descriptions),
  };
}

