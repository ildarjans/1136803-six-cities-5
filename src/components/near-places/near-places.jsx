import React from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";

import {offerPropTypes} from "../../prop-types/offer";

import {PlaceCard} from "../place-card/place-card";
import {selectMemoizeNearOffers} from "../../selectors/offers";

const NearPlacesComponent = ({nearOffers}) => {
  return (
    <div className="container">
      <section className="near-places places">
        <h2 className="near-places__title">Other places in the neighbourhood</h2>
        <div className="near-places__list places__list">
          {nearOffers.map((offer) => (
            <PlaceCard
              key={offer.id}
              offer={offer}
              classNameArticle={`near-places__card`}
              classNameWrapper={`near-places__image-wrapper`}
            />
          ))}
        </div>
      </section>
    </div>
  );
};

NearPlacesComponent.propTypes = {
  nearOffers: PropTypes.arrayOf(offerPropTypes.isRequired).isRequired
};

const mapStateToProps = (state) => ({
  nearOffers: selectMemoizeNearOffers(state)
});

export const NearPlaces = connect(mapStateToProps)(NearPlacesComponent);
