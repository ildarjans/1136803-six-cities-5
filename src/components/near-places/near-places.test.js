import React from "react";
import {Provider} from "react-redux";
import {MemoryRouter} from "react-router-dom";
import renderer from "react-test-renderer";

import {mockCityOffers} from "../../mocks/mock-data";
import {mockStore} from "../../mocks/mock-store";

import {NearPlaces} from "./near-places";

const noop = () => {};
const CARDS_DISPLAY_LIMIT = 3;

it(`Should NearPlaces rendered correctly`, () => {
  const tree = renderer
    .create(
        <Provider store={mockStore}>
          <MemoryRouter>
            <NearPlaces
              id={1}
              nearOffers={mockCityOffers.AMSTERDAM.slice(0, CARDS_DISPLAY_LIMIT)}
              fetchNearOffersById={noop}
            />
          </MemoryRouter>
        </Provider>
    ).toJSON();
  expect(tree).toMatchSnapshot();
});
