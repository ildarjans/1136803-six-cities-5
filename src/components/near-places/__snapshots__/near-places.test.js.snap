// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should NearPlaces rendered correctly 1`] = `
<div
  className="container"
>
  <section
    className="near-places places"
  >
    <h2
      className="near-places__title"
    >
      Other places in the neighbourhood
    </h2>
    <div
      className="near-places__list places__list"
    >
      <article
        className="near-places__card place-card"
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        <div
          className="near-places__image-wrapper place-card__image-wrapper"
        >
          <a
            href="/offer/73"
            onClick={[Function]}
          >
            <img
              alt="Big company apartment"
              className="place-card__image"
              height={200}
              src="http://picsum.photos/248/152?r=0.16844568825898554"
              width={260}
            />
          </a>
        </div>
        <div
          className="place-card__info"
        >
          <div
            className="place-card__price-wrapper"
          >
            <div
              className="place-card__price"
            >
              <b
                className="place-card__price-value"
              >
                €
                2523
              </b>
              <span
                className="place-card__price-text"
              >
                / night
              </span>
            </div>
            <button
              className="place-card__bookmark-button  button"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="place-card__bookmark-icon"
                height={18}
                width={18}
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                To
                 bookmarks
              </span>
            </button>
          </div>
          <div
            className="place-card__rating rating"
          >
            <div
              className="place-card__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "20%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
          </div>
          <h2
            className="place-card__name"
          >
            <a
              href="/offer/73"
              onClick={[Function]}
            >
              Big company apartment
            </a>
          </h2>
          <p
            className="place-card__type"
          >
            apartment
          </p>
        </div>
      </article>
      <article
        className="near-places__card place-card"
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        <div
          className="near-places__image-wrapper place-card__image-wrapper"
        >
          <a
            href="/offer/85"
            onClick={[Function]}
          >
            <img
              alt="Big company apartment"
              className="place-card__image"
              height={200}
              src="http://picsum.photos/248/152?r=0.34977443093807237"
              width={260}
            />
          </a>
        </div>
        <div
          className="place-card__info"
        >
          <div
            className="place-card__price-wrapper"
          >
            <div
              className="place-card__price"
            >
              <b
                className="place-card__price-value"
              >
                €
                3585
              </b>
              <span
                className="place-card__price-text"
              >
                / night
              </span>
            </div>
            <button
              className="place-card__bookmark-button place-card__bookmark-button--active button"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="place-card__bookmark-icon"
                height={18}
                width={18}
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                In
                 bookmarks
              </span>
            </button>
          </div>
          <div
            className="place-card__rating rating"
          >
            <div
              className="place-card__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
          </div>
          <h2
            className="place-card__name"
          >
            <a
              href="/offer/85"
              onClick={[Function]}
            >
              Big company apartment
            </a>
          </h2>
          <p
            className="place-card__type"
          >
            room
          </p>
        </div>
      </article>
    </div>
  </section>
</div>
`;
